verification/article/article.md

```diff
@@ -0,0 +1,1063 @@
+integer arithmetic as $(x \cdot y) / k$. The following key lemmas were proven from first principles:
+
+* **Monotonicity with Favorable Scaling (`Lemma_MulDivGreater_From_Scratch`):** This lemma proves that if the scaling
+  factor is greater than or equal to 1, the result is no less than the original amount.
+
+    * **Property 1.** $\forall x, y, k \in \mathbb{N}$ where $x > 0$, $k > 0$, and $y \ge k$
+      $$\frac{x \cdot y}{k} \ge x$$
+      This is crucial for proving that conversions at a stable or favorable price do not result in a loss of principal.
+
+* **Strict Monotonicity with Highly Favorable Scaling (`Lemma_MulDivStrictlyGreater_From_Scratch`):** Due to integer
+  division truncation, $y > k$ is insufficient to guarantee $(x \cdot y) / k > x$. This lemma establishes a stronger
+  precondition to ensure a strict increase.
+
+    * **Property 2.** $\forall x, y, k \in \mathbb{N}$ where $x > 0$, $k > 0$, and $y \ge 2k$:\
+      $$\frac{x \cdot y}{k} > x$$
+      This is used to prove that a significantly favorable price or a large bonus yields a tangible gain.
+
+* **Round-trip Truncation (`Lemma_DivMul_LTE`):** This lemma formalizes the fundamental property of integer division:
+  information may be lost.
+
+    * **Property 3.** $\forall x, y \in \mathbb{N}$ where $y > 0$:
+      $$ \left( \frac{x}{y} \right) \cdot y \le x $$
+      This property is the cornerstone for proving the safety of round-trip calculations, such as when a refund is
+      processed.
+
+These foundational lemmas abstract away the complexities of integer arithmetic, allowing higher-level modules to reason
+about calculations in terms of simple inequalities.
+
+## 4. Core Business Logic Verification
+
+Building upon the `MathLemmas` foundation, we verify the core business logic components.
+
+### 4.1. Asset Conversion (`AssetCalculations`)
+
+This module defines the logic for converting a base `amount` into assets based on a price fraction
+`saleToken / depositToken`. The specification is:
+`CalculateAssetsSpec(amount, dT, sT) := (amount * sT) / dT`
+
+The module provides lemmas that instantiate the generic mathematical properties for this specific context. For instance,
+`Lemma_CalculateAssets_IsGreaterOrEqual` proves that `CalculateAssetsSpec(...) >= amount` if `sT >= dT`, by directly
+invoking `Lemma_MulDivGreater_From_Scratch`.
+
+A critical property for refund safety is the **round-trip inequality**, proven in
+`Lemma_AssetsRevert_RoundTrip_lte` [@bhargavan2016formal]. It
+states that converting an amount to assets and then immediately converting those assets back to the original currency
+cannot result in a gain.
+
+* **Property 4 (Asset Conversion Round-Trip Safety).** Let $Assets(w) := \text{CalculateAssetsSpec}(w, dT, sT)$ and
+  $Revert(a) := \text{CalculateAssetsRevertSpec}(a, dT, sT)$. Then for $w > 0$:
+  $$ \text{Assets}(w) > 0 \implies \text{Revert}(\text{Assets}(w)) \le w $$
```

Comment:
Possible future work:

As mentioned above, we can use the stronger version of the `DivMul` lemma to bound the losses a user can experience. This may also be useful because we want to guarantee that users are not losing funds in addition to guaranteeing that they cannot create funds from nothing.

For example, the proven property `w > 0 && Assets(w) > 0 -> Revert(Assets(w)) <= w` also holds for a the function `Revert(x) := 0` (i.e. the constant function that always returns 0. Clearly this property is not strong enough to users any confidence their funds are safe.

Reply:
But in was already changed to:
```
* **Property 4 (Asset Conversion Round-Trip Safety and Bounded Loss).**
  Let $Assets(w) := \text{CalculateAssetsSpec}(w, dT, sT)$
  and $Revert(a) := \text{CalculateAssetsRevertSpec}(a, dT, sT)$. Then for $w > 0$:
  $$ \text{Assets}(w) > 0 \implies \text{Revert}(\text{Assets}(w)) \le w $$
  Moreover, the lemma proves a stronger property: the scaled difference between the original and reverted amounts will
  never exceed the sum of the price fraction's terms:
  $$ (w - \text{Revert}(\text{Assets}(w))) \cdot sT < dT + sT $$
  This guarantee is crucial as it proves that financial loss from rounding errors is predictable and has a hard ceiling.

```

and updated **Appendix B.4.3** with new rules. Do you think I should extend more **Property 4**?